Add organization type to org profile settings.

laurynmm opened the pull request:
Adds updating the organization's type to the organization profile admin page. Fixes #21692.

### Back end implementation notes:
- The `Realm.org_type` setting was added in #19042, so there was no need to update the models or do a migration.
- There was a special `RealmAuditLog` event type created for updates to a realm's `org_type`, which has been possible for Zulip Cloud through support admin.
- I extended that implementation, which meant it needed a slightly different handling in `views.realm.update_realm` even though the setting follows the standard structure. Added tests to `test_realm` for this.
- I added an event to `actions.do_change_realm_org_type` so that other users in the realm would get live updates as they do for other organization settings. Added test to `test_events` for this.
- Questions to confirm / consider:
  - Do we want to keep the specific RealmAuditLog tracking for changes to an organization's type as I did here? Or do we want to use the standard update log?
  - When the setting was added, existing realms were updated to be `Unspecified`. Do we want this to be an acceptable value for organizations to update to? If not, do we want to handle on the back end or front end?

### Front end implementation notes:
- This first iteration is a simple drop-down menu with a help center link.
- Questions to confirm / consider:
  - The issue mentions providing some information about how the data will be used. Do we want to do this in the help center documentation or on the organization profile admin page?
    - If on the page, then how / what?
    - If in the help center docs, where /  what?
  - The existing registration and Zulip Cloud support pages do not show the `Unspecified` option. On the one hand, we need that as an option because a large number of existing orgs have that as their `org_type` value. On the other hand, we may not want to let organizations update to this setting? Or maybe that's okay?
  - If we do want to limit this, I'm not sure how to do that with handlebars and in a way that allows us to show the current orgs that are set to `Unspecified`. The other pages are written in html and use Django template syntax to limit the options shown.
  - Do we want to add this information to the realm log-in page, which is where we show the organization's description and profile photo? If so, we would need to update that image used in the help center documentation (see screenshot below). Part of me thinks no or only if the setting isn't set to `Unspecified`.

**Screenshots and screen captures:**

![Screenshot from 2022-04-13 19-33-59](https://user-images.githubusercontent.com/63245456/163244643-75d1c8d6-cf36-4632-bcdd-77925b2fffae.png)

![Screenshot from 2022-04-13 19-34-41](https://user-images.githubusercontent.com/63245456/163244648-f426c011-4294-4923-a8f9-f8d9426b7660.png)

![Screenshot from 2022-04-13 19-55-00](https://user-images.githubusercontent.com/63245456/163244649-f23e3912-ffe2-44be-8313-c955f369a9e7.png)

![Peek 2022-04-13 19-37](https://user-images.githubusercontent.com/63245456/163244813-50b99acf-4969-4130-abc0-3e2aa019f32c.gif)

![Peek 2022-04-13 19-47](https://user-images.githubusercontent.com/63245456/163244832-c91126f9-ab34-4e43-a031-e735b5a80e53.gif)

**Self-review checklist**

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] Self-reviewed the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/version-control.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.

alya commented:
Hm, I'm thinking adding "type" where you did is good, but let's make a separate documentation page for it. New organizations are already going to have the correct org type; we'd like to encourage existing orgs to update theirs.

alya commented:
We can put the new page under Organization basics, as the second link.

The page intro can say something like: "Setting the organization type lets Zulip customize the experience for users in your organization (e.g. Welcome Bot interactions)."

alya commented:
> The existing registration and Zulip Cloud support pages do not show the Unspecified option. On the one hand, we need that as an option because a large number of existing orgs have that as their org_type value. On the other hand, we may not want to let organizations update to this setting? Or maybe that's okay?

Rather than having the menu say "Unspecified", let's have that displayed as the empty string. I think that will be sufficiently discouraging as an option to change your org type to.

alya commented:
> Do we want to add this information to the realm log-in page, which is where we show the organization's description and profile photo?

I don't think we need to show it there.

timabbott commented:
> Rather than having the menu say "Unspecified", let's have that displayed as the empty string. I think that will be sufficiently discouraging as an option to change your org type to.

Yeah, that sounds right to me.

laurynmm commented:
> > Rather than having the menu say "Unspecified", let's have that displayed as the empty string. I think that will be sufficiently discouraging as an option to change your org type to.
> 
> Yeah, that sounds right to me.

So, I don't think an empty string is possible due to the internationalization formatting that we use not allowing empty strings. Would a dash, `-`, work? We don't have any other drop downs in the web app with a blank option. Here's a screenshot of the menu with the dash ...

![Screenshot from 2022-04-17 19-11-43](https://user-images.githubusercontent.com/63245456/163725379-01d5ad71-33e1-464a-b657-5ae714681b72.png)

laurynmm commented:
Updates to PR from front end review comments:
- I've added a separate help center page (see screenshot below). It's currently just the short description and instructions block.
- I did figure out how to have an empty string for the "Unspecified" organization type, but I kind of think it looks better with the dash. Posting some screenshots to the [CZO design](https://chat.zulip.org/#narrow/stream/101-design/topic/communities.20directory.20.2315844/near/1365971) discussion for that.

**Set organization type help documentation page**:
![Screenshot from 2022-04-18 19-49-58](https://user-images.githubusercontent.com/63245456/163851464-e0fbfc14-4277-4c19-b6bd-116b2cfbf039.png)

timabbott commented:
This is still tagged as WIP -- can you comment on what is known to be incomplete here?

laurynmm commented:
> This is still tagged as WIP -- can you comment on what is known to be incomplete here?

There are two pending design discussion / decisions:
1. Where is the drop down in the modal: between the org name and description or after the profile preview link?
2. Do we want a dash or an empty string for the "Unspecified" description?

I didn't get any feedback on the backend questions here yet. The one about the "Unspecified" setting seems to be using the less descriptive text in the UI will be the extent we stop people from choosing that as their organization type. The one about the `RealmAuditLog` type I could post as a question in a topic on CZO backend to see if there's any feedback.

alya commented:
For reference, this is the design thread: https://chat.zulip.org/#narrow/stream/101-design/topic/communities.20directory.20.2315844/near/1367930

@laurynmm as a heads up, it's helpful to have a link to the #design conversation right in the comment for @timabbott to refer to, so that he can easily check whether the questions have been addressed.

laurynmm commented:
@alya Thanks for adding that link! I'll be sure to do that moving forward. Those design questions have been clarified, so I've updated for those decisions and removed the WIP tag.

While writing out my thoughts on the `RealmAuditLog` choice I made in the back end implementation, I feel pretty confident that was the right call. Especially since this is extending the existing implementation used in `analytics.support.py`.

laurynmm commented on zerver/actions/realm_settings.py:
```diff
@@ -403,6 +403,9 @@ def do_change_realm_org_type(
 
+    event = dict(type="realm", op="update", property="org_type", value=org_type)
+    transaction.on_commit(lambda: send_event(realm, event, active_user_ids(realm.id)))
+
```
I kept this decorated as `transaction_atomic(durable=True)` after reading through some [recent conversations in CZO](https://chat.zulip.org/#narrow/stream/3-backend/topic/Choosing.20the.20approach.20for.20atomic.28.29.20in.20do_.2E.2E.2E.2E.20functions/near/1312009), but I thought it was worth noting here to make sure that was the right call.

timabbott replied:
That sounds correct to me.

alya commented:
Cool, after reviewing the PR again, one more point of feedback (suggested by @timabbott ): once you have an option other than "Unspecified" set, the blank line should no longer show up in the drop-down. That is, organizations that entered an option at sign-up will not encounter it, and others will no longer see it after saving their choice.

laurynmm commented:
@alya - I've updated the front end implementation so that the empty string / "Unspecified" dropdown option doesn't appear once an organization has set a different, defined org_type value (see GIF below). I'm not sure if one can create a realm with the org_type as 'Unspecified' for testing in the dev environment. I was doing this manually in the code as I was debugging and testing.

**screen capture of updating an organization with 'unspecified'**:
![Peek 2022-04-21 11-19](https://user-images.githubusercontent.com/63245456/164431766-9c0dad15-4082-480d-9318-a2c3d76c1fb6.gif)

alya commented:
Yeah, I'm not sure how to test on dev, but the screen capture looks good -- thanks!

timabbott commented on static/js/settings_config.ts:
```diff
@@ -441,6 +441,71 @@ export const user_role_values = {
+// Remove the 'unspecified' org_type for dropdown menu options
+// when an org_type other than 'unspecified' has been selected.
+export const defined_org_type_values = Object.fromEntries(
+    Object.entries(all_org_type_values).slice(1),
```
`.slice(1)` is probably more dependant on the order of the entries for its correctness than would be ideal.

timabbott replied:
I didn't fix this since it seemed a little fiddly, but it'd be a good small follow-up PR.

timabbott commented:
This looks great; I applied the following edits:
* Renamed /help/ article to "Organization type" and extended its content a bit to make it easier to talk about other things this setting does. I think the main motivation was that the common question is likely "what does organization type do", not "How do I change my organization type", since the vast majority of organizations likely will only set it once during account creation.
* Added a bit more detail to the /api/ docs to reassure client developers that they can probably ignore this, and of course like to the /help/ article.

```
diff --git a/static/templates/settings/organization_profile_admin.hbs b/static/templates/settings/organization_profile_admin.hbs
index e9c240bdf1..66c141925b 100644
--- a/static/templates/settings/organization_profile_admin.hbs
+++ b/static/templates/settings/organization_profile_admin.hbs
@@ -19,7 +19,7 @@
                 </div>
                 <div class="input-group admin-realm">
                     <label for="realm_org_type" class="dropdown-title">{{t "Organization type" }}
-                        {{> ../help_link_widget link="/help/set-organization-type" }}
+                        {{> ../help_link_widget link="/help/organization-type" }}
                     </label>
                     <select name="realm_org_type" class="setting-widget prop-element" id="id_realm_org_type" data-setting-widget-type="number">
                         {{> dropdown_options_widget option_values=realm_org_type_values}}
diff --git a/templates/zerver/help/include/sidebar_index.md b/templates/zerver/help/include/sidebar_index.md
index 7162233142..7640e43a43 100755
--- a/templates/zerver/help/include/sidebar_index.md
+++ b/templates/zerver/help/include/sidebar_index.md
@@ -128,7 +128,7 @@
 
 ## Organization basics
 * [Review your organization's settings](/help/review-your-organization-settings)
-* [Set organization type](/help/set-organization-type)
+* [Organization type](/help/organization-type)
 * [Import from Mattermost](/help/import-from-mattermost)
 * [Import from Slack](/help/import-from-slack)
 * [Import from Gitter](/help/import-from-gitter)
diff --git a/templates/zerver/help/organization-type.md b/templates/zerver/help/organization-type.md
new file mode 100644
index 0000000000..4f8c6ddf52
--- /dev/null
+++ b/templates/zerver/help/organization-type.md
@@ -0,0 +1,24 @@
+# Organization type
+
+An organization's type is used to customize the experience for users
+in your organization, including initial organization settings and
+Welcome Bot messages received by new users.
+
+## Set organization type
+
+{!admin-only.md!}
+
+{start_tabs}
+
+{settings_tab|organization-profile}
+
+1. Under **Organization type**, select the option that best fits
+your organization.
+
+{!save-changes.md!}
+
+{end_tabs}
+
+## Related articles
+
+* [Configure default new user settings](/help/configure-default-new-user-settings)
diff --git a/templates/zerver/help/set-organization-type.md b/templates/zerver/help/set-organization-type.md
deleted file mode 100644
index aa9fc8e8ed..0000000000
--- a/templates/zerver/help/set-organization-type.md
+++ /dev/null
@@ -1,19 +0,0 @@
-# Set organization type
-
-{!admin-only.md!}
-
-Setting the organization type lets Zulip customize the experience
-for users in your organization (e.g. Welcome Bot interactions).
-
-## Set organization type
-
-{start_tabs}
-
-{settings_tab|organization-profile}
-
-1. Under **Organization type**, select the option that best fits
-your organization.
-
-{!save-changes.md!}
-
-{end_tabs}
diff --git a/zerver/openapi/zulip.yaml b/zerver/openapi/zulip.yaml
index b7ec0ca814..9a6d1fdadb 100644
--- a/zerver/openapi/zulip.yaml
+++ b/zerver/openapi/zulip.yaml
@@ -3932,7 +3932,8 @@ paths:
                                     org_type:
                                       type: integer
                                       description: |
-                                        The type of the organization.
+                                        The [organization type](/help/organization-type)
+                                        for the realm.
 
                                         - 0 = Unspecified
                                         - 10 = Business
@@ -11488,7 +11489,10 @@ paths:
                         description: |
                           Present if `realm` is present in `fetch_event_types`.
 
-                          The type of the organization.
+                          The [organization type](/help/organization-type) for the realm.
+                          Useful only to clients supporting changing this setting for the
+                          organization, or clients implementing onboarding content or
+                          other features that varies with organization type.
 
                           - 0 = Unspecified
                           - 10 = Business
```

timabbott commented:
I've tagged this to merge once CI finishes.
